Sources/CoreMetrics/SystemMetrics.swift

```diff
@@ -0,0 +1,235 @@
+        self.prefix = prefix
+        self.virtualMemoryBytes = virtualMemoryBytes
+        self.residentMemoryBytes = residentMemoryBytes
+        self.startTimeSeconds = startTimeSeconds
+        self.cpuSecondsTotal = cpuSecondsTotal
+        self.maxFds = maxFds
+        self.openFds = openFds
+    }
+
+    func label(for keyPath: KeyPath<SystemMetricsLabels, String>) -> String {
+        return self.prefix + self[keyPath: keyPath]
+    }
+}
+
+public protocol SystemMetrics {
+    init(pid: Int)
+
+    var virtualMemoryBytes: Int? { get }
+    var residentMemoryBytes: Int? { get }
+    var startTimeSeconds: Int? { get }
+    var cpuSeconds: Int? { get }
+    var maxFds: Int? { get }
+    var openFds: Int? { get }
+}
+
+#if os(Linux)
+internal struct LinuxSystemMetrics: SystemMetrics {
+    let virtualMemoryBytes: Int?
+    let residentMemoryBytes: Int?
+    let startTimeSeconds: Int?
+    let cpuSeconds: Int?
+    let maxFds: Int?
+    let openFds: Int?
+
+    private enum SystemMetricsError: Error {
+        case MetricReadError
+    }
+    
+    init(pid: Int) {
+        let pid = "\(pid)"
+        let ticks = _SC_CLK_TCK
+        
+        do {
+            guard
+                let fp = fopen("/proc/\(pid)/stat", "r")
+            else { throw SystemMetricsError.MetricReadError }
+            var buf = [CChar](repeating: CChar(0), count: 1024)
+            while fgets(&buf, 1024, fp) != nil { }
+            guard fclose(fp) == 0 else { throw SystemMetricsError.MetricReadError }
+            guard
+                let statString = String(cString: buf)
+                .split(separator: ")")
+                .last
+            else { throw SystemMetricsError.MetricReadError }
+            let stats = String(statString)
+                .split(separator: " ")
+                .map(String.init)
+            self.virtualMemoryBytes = Int(stats[safe: 20])
+            if let rss = Int(stats[safe: 21]) {
+                self.residentMemoryBytes = rss * _SC_PAGESIZE
+            } else {
+                self.residentMemoryBytes = nil
+            }
+            if let startTimeTicks = Int(stats[safe: 19]) {
+                self.startTimeSeconds = startTimeTicks / ticks
+            } else {
+                self.startTimeSeconds = nil
+            }
+            if let utimeTicks = Int(stats[safe: 11]), let stimeTicks = Int(stats[safe: 12]) {
+                let utime = utimeTicks / ticks, stime = stimeTicks / ticks
+                self.cpuSeconds = utime + stime
+            } else {
+                self.cpuSeconds = nil
+            }
+        } catch {
+            self.virtualMemoryBytes = nil
+            self.residentMemoryBytes = nil
+            self.startTimeSeconds = nil
+            self.cpuSeconds = nil
+        }
+        do {
+            var _rlim = rlimit()
+            
+            guard getrlimit(__rlimit_resource_t(RLIMIT_NOFILE.rawValue), &_rlim) == 0 else { throw SystemMetricsError.MetricReadError }
```

Comment:
here as well, please swap `&_rlim` for `with*Unsafe*(&_rlim) { rLimPtr in ... }`